File: TypeScript/src/trip/TripService.ts
import UserNotLoggedInException from "../exception/UserNotLoggedInException";
import User from "../user/User";
import { UserSession } from "../user/UserSession";
import Trip from "./Trip";
import TripDAO from "./TripDAO";

export default class TripService {
  constructor(private userSession: UserSession, private tripDAO: TripDAO) {}

  public getTripsByUser(user: User): Trip[] {
    const loggedUser = this.getLoggedUsers();

    if (loggedUser == null) {
      throw new UserNotLoggedInException();
    }

    return this.getTripList(user, loggedUser);
  }

  protected findUserTrips(user: User): Trip[] {
    return this.tripDAO.findTripsByUser(user);
  }

  protected getLoggedUsers(): User {
    return this.userSession.getLoggedUser();
  }

  private getTripList(user: User, loggedUser: User) {
    const isFriend = user.isFriend(loggedUser);
    if (isFriend) {
      return this.findUserTrips(user);
    }
    return [];
  }
}

File: scala/trip_service_kata/src/main/scala/org/craftedsw/tripservicekata/trip/TripService.scala
package org.craftedsw.tripservicekata.trip

import org.craftedsw.tripservicekata.exception.UserNotLoggedInException
import org.craftedsw.tripservicekata.user.{Session, User}

import scala.util.control.Breaks._

class TripService(tripDao: TripDAOTrait, session: Session) {

  def getTripsByUser(myUser: User): List[Trip] = {
      session getLoggedUser() match {
      case None => throw new UserNotLoggedInException
      case Some(anotherUser) => getTripsCheckingFriendship(myUser, anotherUser)
    }
  }

  private def getTripsCheckingFriendship(myUser: User, anotherUser: User) = {
    if (myUser.isFriendOf(anotherUser)) tripDao.findTripsByUser(myUser)
    else List()
  }
}
